### DISCUSSION TITLE
How to `assert_never` in a `match` statement

### DISCUSSION THREAD

**Discussion by Jasha10** [microsoft/pyright#5186] [General]
I'd like to call `assert_never` in the irrefutable branch of a `match` statement:

```python
from typing_extensions import assert_never

def get_val() -> int | str:
    return 1

match get_val():
    case int():
        print("got int")
    case str():
        print("got str")
    case never:
        assert_never(never)
```

Calling `assert_never` like this (as opposed to omitting the final match case) has some benefits:
- It makes clear to humans that all possibilities are exhausted by the preceding cases, and
- It forces type checkers to verify that `case never` won't be reached.
- It gives an `AssertionError` at runtime if the case *is* reached due to a bug.

Pyright complains about the above usage:
```bash
$ pyright tmp.py
/home/homestar/dev/mrepo_wip/pysc/tmp.py
  /home/homestar/dev/mrepo_wip/pysc/tmp.py:11:10 - error: Pattern will never be matched for subject type "Never" (reportUnnecessaryComparison)
1 error, 0 warnings, 0 informations
$ pyright --version
pyright 1.1.310
```
My solution for now is to ignore this pyright error:
```python
...
    case never:  # pyright: ignore [reportUnnecessaryComparison]
        assert_never(never)
```
Is there a better way to assert that the match arms are exhaustive (i.e. one that doesn't trigger a pyright error)? If not, @msfterictraut, would you be amenable to a feature request to ease pyright's error reporting for this use-case?

**Comment by Jasha10** [comment]
I found that the Python typing team has published example code that endorses use of `assert_never` in the last arm of a match statement:
https://typing.readthedocs.io/en/latest/source/unreachable.html#assert-never-and-exhaustiveness-checking
Given the official endorsement of this pattern, I will close this discussion and instead open a feature request.

Edit:
So it turns out that pyright only rejects this pattern if `typeCheckingMode` is set to `"strict"`. Perhaps I was too quick to open this discussion. Mea culpa.

**Reply by erictraut** [reply]
This is working as intended. Pyright already provides a way to detect non-exhaustive matches using the `reportMatchNotExhausted`, so there's no need to use `assert_never`.

If you include the unnecessary case statement and enable the `reportUnnecessaryComparison` rule, pyright will tell you that the case statement isn't needed.

So your options are:
1. Enable `reportMatchNotExhausted` (which is on by default in strict mode) and do not add the extraneous case statement. (This is the solution I recommend.)
2. Leave the extraneous case statement in place and do one of the following:
- Disable the `reportUnnecessaryComparison` rule (which is off by default in "basic" type checking mode but on by default in "strict" mode.
- Use a `# type: ignore` or `# pyright: ignore` comment to suppress the error.
